typedconfig/casts.py

```diff
@@ -1,14 +1,111 @@
 from enum import Enum
-from typing import TypeVar, Callable, Type
+from typing import Callable, List, Optional, Type, TypeVar, Union, overload
 
+T = TypeVar("T")
 TEnum = TypeVar("TEnum", bound="Enum")
 
 
 def enum_cast(enum_type: Type[TEnum]) -> Callable[[str], TEnum]:
+    """
+    A cast function that translates a string input into a member of named Enum
+    Throws a KeyError of the string is not in the given Enum
+
+    Parameters
+    ----------
+    enum_type
+       The EnumType that the string input should be cast to
+
+    Returns
+    -------
+    Cast method that can be used with the key() function
+    """
+
     def getter(name: str) -> TEnum:
         try:
             return enum_type[name]
         except KeyError:
             raise KeyError(f"{name} is not a member of {enum_type}")
 
     return getter
+
+
+@overload
+def list_cast(
+    base_cast: None = ...,
+    delimiter: str = ...,
+    ignore_trailing_delimiter: bool = ...,
+    strip: bool = ...,
+) -> Callable[[str], List[str]]:
+    ...
+
+
+@overload
+def list_cast(
+    base_cast: Optional[Callable[[str], T]] = ...,
```

Comment:
I think this line should just be
```python
base_cast: Callable[[str, T]],
```

I'm not sure I'm 100% clear on the nuances of `@overload`, but as far as I understand, these overloads are trying to deal with the two cases:
(a) User does not provide a value for `base_cast` or provides a value for `base_cast` of `None`. In this case we will be casting to `List[str]`
(b) User provides a function to `base_cast`. In this case we will be casting to `List[T]`

In case (b), therefore, a value is provided for `base_cast` and therefore the overload should not include the `= ...` which indicates that the input is optional. Similarly the `Optional` is not required since in this case the user is not providing a value of `None` since that is handled by case (a).

Without this change, the type inference works correctly in mypy but PyCharm's autocomplete does not pick it up. With this change, both work.

Reply:
Good catch on the @overload. I'll fix that. 

And I'll change the whole thing to be tuple_cast. I did think of that - it's a bit odd to have a mutable type inside an immutable structure. There could be a use for other immutable collection types, like a frozen_set_cast, but I'd stick to just tuple_cast for now.